### PR TITLE
Include status when returning response from requests to mavenlink

### PR DESCRIPTION
**Pull request by azellsworth**
This proposed change would include the status in the response of any request from Mavenlink. The main purpose of this change is to [use the status code to determine the type of error on our integrations platform](https://github.com/mavenlink-solutions/konekti/pull/984).  

For most requests (GET, POST, PUT) we are just adding a status code to the hash that's returned from the gem. However, for DELETE, the response has historically been a blank hash. Adding in the status code would yield a hash with just a status code. Thus, if any users of the gem are checking for `response.blank?` on a delete request, we will be introducing a breaking change. 

We would appreciate comments as to whether this is an appropriate change to the gem. 

For the time being, we are able to look for specific error messages on the integrations platform, so this is a desired change but not required for us to properly handle the errors. We would prefer to use the error code so that we do not have to couple the integrations platform to the Mavenlink API around specific error messages.

### CODE REVIEW

File: lib/mavenlink/client.rb [hunk 1]
```diff
@@ -87,16 +91,16 @@ def connection_options
       }.freeze
     end
 
-    def parse_request(response)
-      if response.present?
-        parsed_response = JSON.parse(response)
+    def parse_request(body, status)
```
Comment (azellsworth) [inline lib/mavenlink/client.rb]: what if response doesn't have a status?

Comment (diclophis) [reply lib/mavenlink/client.rb]: A response is almost certainly guaranteed to have a `.status` if it has a `.body`

File: lib/mavenlink/client.rb [hunk 2]
```diff
@@ -87,16 +91,16 @@ def connection_options
...
+      if body.present?
+        parsed_response = JSON.parse(body).merge("status" => status)
       else
         return
       end
 
       parsed_response.tap do
         Mavenlink.logger.whisper 'Received response:'
-        Mavenlink.logger.inspection response
+        Mavenlink.logger.inspection body
```
Comment (diclophis) [inline lib/mavenlink/client.rb]: This could be a secure information leak if these log messages are beamed off the box

**Comment by diclophis** [comment]
I think generally we would prefer the "higher order" abstraction of this gem to provide a simple abstraction of

```
post = Mavenlink::Post.find(1)
post.destroy # returns boolean only, setting post.errors to some values of what failed, or raising an exception on failure/5xx response
```

If anything we should provide a way to tap into the request/response cycle to allow users of this gem to custom handle logging of the specifics of a given request ... in general the design of this gem is to "mirror" ActiveRecord like invocations... so in that sense, the specifics of the http response code should _never_ be visible at the "surface" interface that this gem exposes...

Long story short, if we need better logging of response codes, we should provide a callback mechanism so we can track that.

**Comment by ahuth** [comment]
@diclophis are you thinking some sort of API like this:

```rb
mavenlink_gem.get('foo', params) do |response|
  if response.status == 666
    # Stuff
  end
end
```

which end uses can use, without needing to make assumptions about what information is needed to be returned?

**Comment by diclophis** [comment]
@ahuth if we want that style i think we should make a different client... this gem is currently programmed to behave as much like activerecord as possible ... i think anyways? thats the way the README reads

File: lib/mavenlink/client.rb (hunk 2 above)
Comment (jacksenechal) [reply lib/mavenlink/client.rb]: Agreed. That sounds like an existing issue with the gem. We'll definitely want to consider the secure data issue when we implement a separate logging service. That comes a bit later in the roadmap.

File: lib/mavenlink/client.rb (hunk 1 above)
Comment (jacksenechal) [reply lib/mavenlink/client.rb]: 👍

**Comment by ahuth** [comment]
Ah, that style is what I imagined by your suggestion (which is a good one, I think) to allow users to tap into the request lifecycle.

Any thoughts on what that would look like?

**Comment by diclophis** [comment]
Maybe something like this:

```
# global hook

Mavenlink.client.on_http_done do |all_the_http_bits|
  # do stuff like Rails.logger.warn(all_http_bits.inspect)
end

# example per model hook

Mavenlink::Workspace.on_http_done do |all_the_http_bits|
  # do stuff like Rails.logger.warn(all_http_bits.inspect)
end
```

**Comment by ahuth** [comment]
Gotcha, that makes sense. So more of a 'configuration' thing than an actual runtime, in the code, thing

**Comment by diclophis** [comment]
Yea, the primary gist being... why would a user of this particular gem ever want to know anything about anything HTTP related... thats "low level" from this perspective (and for another view point, imagine in the future where the "backend" of this gem _isnt_ http at all (perhaps grpc,websockets,et.al.)).... but i do see a need where a user might want to "introspect" into the mechanics inorder to understand things better, and enable "wiretracing" on `$whateverBackInterface` exists at the given time.
